Adopting Structured Fields (#1297)
httpwg/http-extensions: draft-ietf-httpbis-message-signatures.md

## HTTP Header Fields

> An HTTP header field is identified by its header field name.  While HTTP header field names are case-insensitive, implementations MUST use lowercased field names (e.g., `content-type`, `date`, `etag`) when using them as content identifiers. 

Reviewer: ```suggestion
An HTTP header field is identified by its header field name.  While HTTP header field names are case-insensitive, implementations MUST use lowercased field names (e.g., `content-type`, `date`, `etag`) when using them as content identifiers.
```

|Header Field|Canonicalized Value|
|--- |--- |
|`cache-control`|max-age=60, must-revalidate|
|`date`|Tue, 07 Jun 2014 20:51:35 GMT|
|`server`|www.example.com|
|`x-empty-header`||
|`x-obs-fold-header`|Obsolete line folding.|
|`x-ows-header`|Leading and trailing whitespace.|
{: title="Non-normative examples of header field canonicalization."}

> ## Dictionary Structured Field Members

Reviewer 2: this PR does not specify the actual format of a dictionary

Reviewer 2: disregard - wasn't aware of raft-kamp-httpbis-structure-01

Author: FYI, current WG version of the Structured Fields draft is here: https://tools.ietf.org/html/draft-ietf-httpbis-header-structure

> An individual member in the value of a Dictionary Structured Field is identified by the lowercased field name, followed by a semicolon `":"`, followed by the member name. An individual member in the value of a Dictionary Structured Field is canonicalized by applying the serialization algorithm described in Section 4.1.2 of {{StructuredFields}} on a Dictionary containing only that member.

Reviewer: ```suggestion
An individual member in the value of a Dictionary Structured Field is identified concatenating
the lowercased field name,
a semicolon `":"`
and the member name.
An individual member in the value of a Dictionary Structured Field is canonicalized by applying the serialization algorithm described in Section 4.1.2 of {{StructuredFields}} on a Dictionary containing only that member.
```

Reviewer: @richanna following the discussion at the httpwg meeting I suggest to split this part in another PR so that we can land this one.

Sub-referencing header could be leveraged to push the reicever to parse a given header in a certain way which is in control of the sender (eg. triggering keyerror, indexerror, ..).

Reviewer 2: There is also a typo - semicolon/colon

A prefix of a List Structured Field consisting of the first N members in the field's value (where N is an integer greater than 0 and less than or equal to the number of members in the List) is identified by the lowercased field name, followed by a semicolon `":"`, followed by N expressed as an Integer String. A list prefix is canonicalized by applying the serialization algorithm described in Section 4.1.1 of {{StructuredFields}} on a List containing only the first N members as specified in the list prefix, in the order they appear in the original List.

### Canonicalization Examples

This section contains non-normative examples of canonicalized values for list prefixes given the following example header fields, whose values are assumed to be Dictionaries:

~~~
X-List-A: (a, b, c, d, e, f)
X-List-B: ()
~~~

The following table shows example canonicalized values for different content identifiers, given those fields:

|Content Identifier|Canonicalized Value|
|--- |--- |
|`x-list-a:0`|()|
|`x-list-a:1`|(a)|
|`x-list-a:3`|(a, b, c)|
|`x-list-a:6`|(a, b, c, d, e, f)|
|`x-list-b:0`|()|
{: title="Non-normative examples of list prefix canonicalization."}

## Signature Creation Time

> The signature's Creation Time ({{signature-metadata}}) is identified by the `*created` identifier.

Reviewer: I like the parentheses removal!

Reviewer 2: related: #1303

alg
: 
: RECOMMENDED. The `alg` parameter is a Token containing the name of the signature's Algorithm, as registered in the HTTP Signature Algorithms Registry defined by this document. Verifiers MUST determine the signature's Algorithm from the `keyId` parameter rather than from `alg`. If `alg` is provided and differs from or is incompatible with the algorithm or key material identified by `keyId` (for example, `alg` has a value of `rsa-sha256` but `keyId` identifies an EdDSA key), then implementations MUST produce an error.

created
: 
: RECOMMENDED. The `created` parameter is a Decimal containing the signature's Creation Time, expressed as the canonicalized value of the `*created` content identifier, as defined in Section 2.  If not specified, the signature's Creation Time is undefined.  This parameter is useful when signers are not capable of controlling the Date HTTP Header such as when operating in certain web browser environments.

expires
: 

> : OPTIONAL. The `expires` parameter is a Decimal containing the signature's Expiration Time, expressed as the canonicalized value of the `*expires` content identifier, as defined in Section 2.  If the signature does not have an Expiration Time, this parameter MUST be omitted.  If not specified, the signature's Expiration Time is undefined.

Reviewer: ```suggestion
: OPTIONAL. The `expires` parameter is a Decimal containing the signature's Expiration Time, expressed as the canonicalized value of the `*expires` content identifier, as defined in Section 2.
If the signature does not have an Expiration Time, this parameter MUST be omitted.  If not specified, the signature's Expiration Time is undefined.
```

Author: Both statements are correct.
1. When generating a `Signature-Input` member, if the signature's Expiration Time is undefined, you MUST NOT include the `*expires` parameter.
2. When parsing a `Signature-Input` member, if the member does not have an `*expires` parameter, that means the Expiration Time is undefined.

Reviewer: I would replace the current statement with your (1).

Hint: it would be great to have the statements on different lines like in the above suggestion. Readability and Referencing issues would be simpler especially for non-native speakers :P

    alg=hs2019; created=1402170695; expires=1402170995
Signature: sig1=:K2qGT5srn2OGbOIDzQ6kYT+ruaycnDAAUpKv+ePFfD0RAxn/1BUe
    Zx/Kdrq32DrfakQ6bPsvB9aqZqognNT6be4olHROIkeV879RrsrObury8L9SCEibe
    oHyqU/yCjphSmEdd7WD+zrchK57quskKwRefy2iEC5S2uAH0EPyOZKWlvbKmKu5q4
    CaB8X/I5/+HLZLGvDiezqi6/7p2Gngf5hwZ0lSdy39vyNMaaAT0tKo6nuVw0S1MVg
    1Q7MpWYZs0soHjttq0uLIA3DIbQfLiIvK6/l0BdWTU7+2uQj7lBkQAsFZHoA96ZZg
    FquQrXRlmYOh+Hx5D9fJkXcXe5tmAg==:
~~~

Since `Signature-Input` and `Signature` are both defined as Dictionary Structured Headers, they can be used to easily include multiple signatures within the same HTTP message. For example, a signer may include multiple signatures signing the same content with different keys and/or algorithms to support verifiers with different capabilities, or a reverse proxy may include information about the client in header fields when forwarding the request to a service host, and may also include a signature over those fields and the client's signature. The following is a non-normative example of header fields a reverse proxy might add to a forwarded request that contains the signature in the above example:

~~~
X-Forwarded-For: 192.0.2.123

> Signature-Input: reverse_proxy_sig=(*created, host, date,

Reviewer: I like `Signature-Input`! Do you think it would be useful to separate the signed headers and signature parameters (*alg, *created, *expires, ...)?

Reviewer 3: That's already done through the use of the structured elements and being able to "name" the signature input fields.

Author: I think we can actually replace content identifiers for signature metadata with signing the whole Signature-Input member (e.g., `signature-input:reverse_proxy_sig`). What remains then is header fields and identifiers for other protocol elements like `*request-target` (and similar likely-to-be-defined identifiers like `*method`, `*path`, etc.). I think splitting those out would add complexity without any clear benefit. It would also complicate the field format, since a dictionary member can't contain two lists.
